accepted/00000-ptf-repos-in-uyuni-2.md

```diff
@@ -0,0 +1,179 @@
+ptf master package via yast or "zypper in" works, but they
+can not be pulled in via dependencies.
+
+Each individual rpm package must require the specific
+master ptf package:
+
+    Name: pkg1
+    Requires: ptf-1234 = 1-1
+
+This makes sure that the solver cannot pull in the rpm
+packages, as that would mean to also pull in the blacklisted
+master ptf package.
+
+
+
+### Updates
+
+Ptfs can be updated to a new release by calling 'zypper up'.
+This will update the master ptf package to some higher
+version (e.g. ptf-1234 = 2-1) and also pull in the corresponding
+rpm packages.
+
+This should work with any installer.
+
+### Uninstalling ptfs
+
+'zypper rm ptf-1234' will uninstall the ptf and revert back to
+non-ptf packages.
+
+We need to check if this is working out of the box or if we need to
+implement something in Uyuni to make it work.
+
+### Making sure that no fixes are lost
+
+If all of the bugs fixed by a ptf are also fixed in maintenance
+updates, a new ptf release consisting only of the master ptf
+package is done:
+
+    Name: ptf-1234
+    Version: 3
+    Release: 1                 # always 1
+    Provides: self-destruct-pkg()
+    Requires: (pkg1 >= maintpkg1EVR if pkg1)
+    Requires: (pkg2 >= maintpkg2EVR if pkg2)
+
+Updating to this version will make sure that this system will
+contain only the fixed packages from the maintenance updates.
+The special "self-destruct-pkg()" provides will tell the
+solver that this will be a package erase instead of an
+install. This means that installing this package will actually
+erase the master ptf package.
+
+### Test Packages
+
+Before generating PTFs, SUSE will first create so called TEST packages.
+The customer need to test them and report positive result before SUSE make an official supported PTF out of it.
+
+TEST packages should only be installed on dedicated test systems.
+Therefor it is important that they get not installed on any system where just the version of the TEST package is higher than the installed one.
+
+Current plan is to ship test packages in separate channels.
+
+
+## PTFs with Uyuni for SUSE and non SUSE OSes
+
+The general PTF design has 3 key points.
+
+1. A new ptf package to install the PTF.
+   
+   The "boolean dependencies" are supported in `rpm` since version 4.13.
+   This would work for SLE15+ and RHEL8+.
+   SLE12 will get a different solution with some enhancements to the solver.
+   
+   - RHEL 8 and Clones use an rpm version which provide the required syntax.
+   - RHEL 7 and Clones cannot use `Requires`, they should use `Conflicts` with the limitation of not supporting multiversion installations.
+   - Ubuntu and Debian do not support something like the `if installed` syntax for dependencies, but the `Conflicts` trick should work.
+   
+   The number of packages which potentially get a PTF for Non-SUSE OSes is very limited. We only ship PTF packages for the Client Tools which
+   is a handlul of packages.
```

Comment:
Is it also correct to say that none of the "PTF-able" packages for non-SUSE OSs are multi-versioned (currently and in the foreseeable future)?

Just to double check if I got it right: the only problematic OS is SLES-ES 7, as we could have multiversioned packages there we might produce a PTF for. What would happen in case we have a PTF for, say, SLES-ES 7's kernel?